tests/jsmarionette/client/marionette-client/test/marionette/drivers/promises-test.js

```diff
@@ -0,0 +1,97 @@
+/**
+ * Created by anatal on 12/10/15.
+ */
+/* global assert, helper */
+'use strict';
+suite('marionette/drivers/promises', function() {
+
+  if (typeof(window) !== 'undefined') {
+    return;
+  }
+
+  var Driver = require('../../../lib/marionette/drivers/promises');
+  var net = require('net');
+
+  function issueFirstResponse(subject) {
+    subject.tcp._onDeviceResponse({
+      id: subject.tcp.connectionId,
+      response: {}
+    });
+  }
+
+  setup(function() {
+  });
+
+  teardown(function() {
+  });
+
+  test('should return a fulfilled promise on connect', function(done) {
+    var port = 60769;
+    var subject = new Driver({ port: port });
+
+    var _promise = subject.connect(function(){
+    });
+
+    setTimeout(function() {
+      var server = net.createServer(function(socket) {
+        issueFirstResponse(subject);
+      }).listen(port);
+    }, 50);
+
+    _promise.then(
+      function onFulfill(){
+        done();
+      },
+      function onReject(aRejectReason) {
+      }
+    );
+
+  });
+
+  test('should send an object and receive a promise', function(done) {
+    var port = 60869;
+    var subject = new Driver({ port: port });
+    var sentobj = {type: 'foo'};
+
+    setTimeout(function() {
+
+      var _promiseconnection = subject.connect(function(){
+      });
+
+      setTimeout(function() {
+        var server = net.createServer(function(socket) {
```

Comment:
I was doing this way to use different ports for each socket, but I changed it back to create them in the setup.  

The reason is that sometimes the second test fails to bind because the previous didn't release it yet and we got a 'EADDRINUSE'.

It's intermittent but can happen. See the second run on this log for example:

mozillas-MacBook-Pro-8:marionette-client anatal$ ../../../../node_modules/.bin/mocha  --require test/helper.js --ui tdd --recursive --reporter spec test/marionette/drivers/promises-test 


  marionette/drivers/promises
    ✓ should return a fulfilled promise on connect (102ms)
    ✓ should send an object and receive a promise (39ms)


  2 passing (233ms)

mozillas-MacBook-Pro-8:marionette-client anatal$ ../../../../node_modules/.bin/mocha  --require test/helper.js --ui tdd --recursive --reporter spec test/marionette/drivers/promises-test 


  marionette/drivers/promises
    ✓ should return a fulfilled promise on connect (320ms)
    1) should send an object and receive a promise


  1 passing (465ms)
  1 failing

  1) marionette/drivers/promises should send an object and receive a promise:
     Uncaught Error: listen EADDRINUSE :::60869
      at Object.exports._errnoException (util.js:874:11)
      at exports._exceptionWithHostPort (util.js:897:20)
      at Server._listen2 (net.js:1234:14)
      at listen (net.js:1270:10)
      at Server.listen (net.js:1366:5)
      at null._onTimeout (test/marionette/drivers/promises-test.js:33:10)



mozillas-MacBook-Pro-8:marionette-client anatal$ ../../../../node_modules/.bin/mocha  --require test/helper.js --ui tdd --recursive --reporter spec test/marionette/drivers/promises-test 


  marionette/drivers/promises
    ✓ should return a fulfilled promise on connect (317ms)
    ✓ should send an object and receive a promise


  2 passing (425ms)

mozillas-MacBook-Pro-8:marionette-client anatal$ ../../../../node_modules/.bin/mocha  --require test/helper.js --ui tdd --recursive --reporter spec test/marionette/drivers/promises-test 


  marionette/drivers/promises
    ✓ should return a fulfilled promise on connect (298ms)
    ✓ should send an object and receive a promise


  2 passing (369ms)

mozillas-MacBook-Pro-8:marionette-client anatal$ ../../../../node_modules/.bin/mocha  --require test/helper.js --ui tdd --recursive --reporter spec test/marionette/drivers/promises-test 


  marionette/drivers/promises
    ✓ should return a fulfilled promise on connect (270ms)
    ✓ should send an object and receive a promise


  2 passing (297ms)